### PR TITLE
Minor changes to tincr checkpoints for FPL

### PR DESCRIPTION
**Pull request by ttown523**
## In this commit:
- Updated RapidSmith checkpoint formats
- Fixed a bug with a "get_part" function call that is no longer supported by Vivado
- Added a -verbose option when importing a TCP checkpoint, and added a function to print verbosely.
- Replaced several places that had tabs, with spaces
- modified the genCellLibrary.tcl to only print out leaf cell information (i.e. skip macros for now)

### CODE REVIEW

File: tincr/io/design/tincr_checkpoints.tcl [hunk 1]
```diff
@@ -520,7 +521,7 @@ proc ::tincr::write_routing_rs2 {args} {
 			if {[llength $tiles] == 2} {
 				# assuming that the second tile in the tile list is the interconnect tile
 				set switchbox_tile [lindex $tiles 1]
-				set route_string [string range [get_property ROUTE $net] 3 end-3]
+				set route_string [string range [get_property ROUTE $net] 3 end-4]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Add a comment explaining what this line of code is doing (perhaps even include an example)

File: tincr/io/design/tincr_checkpoints.tcl [hunk 2]
```diff
@@ -223,14 +223,15 @@ proc ::tincr::read_tcp {args} {
...
 	}
 	
-	
 	# unlock the design at the end of the import process...do we want to do this?
-	puts "Design importation complete. ($total_runtime seconds)"
-	remove_files -quiet *
+	if {!$quiet} {
+		puts "Design importation complete. ($total_runtime seconds)"
+	}
+	# remove_files -quiet *
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Why are you commenting out the remove_files command?

Comment (ttown523) [reply tincr/io/design/tincr_checkpoints.tcl]: I added the -no_add option with the read_xdc commands. I uncommented this so all files are now removed. Keep in mind, that this file is still a work in progress, and so there are several places that I have commented out for testing different things, but don't want to delete.

**Comment by ttown523** [comment]
In the latest commit, I refactored the "read_tcp" function a bit to clean some things up. Also, I added a few comments to help improve readability. There are still few changes in this pull request.

File: tincr/cad/util/util.tcl [hunk 3]
```diff
@@ -51,7 +51,9 @@ namespace eval ::tincr {
         process_handler \
         spawn_vivado_run \
         run_in_temporary_project \
-        organize_by
+        organize_by \
+		print_quiet \
```
Comment (bradselw) [inline tincr/cad/util/util.tcl:55]: Spaces instead of tabs

File: tincr/cad/util/util.tcl [hunk 4]
```diff
@@ -1337,6 +1339,18 @@ proc ::tincr::catch_info { cmd } {
 #  namespace import ::tincr::*
 interp alias {} ::tincr::import_all {} namespace import ::tincr::*
 
+proc ::tincr::print_quiet { message quiet {newline ""} } {
+	if {!$quiet} {
+		puts $newline $message
+	}
+}
+
+proc ::tincr::time_command { command } {
```
Comment (bradselw) [inline tincr/cad/util/util.tcl]: Tincr's `report_runtime` command already returns the runtime of any given command (in various formats, including seconds). Remove `time_command` and use `report_runtime <command> s` instead.

File: tincr/cad/util/util.tcl [hunk 5]
```diff
@@ -1337,6 +1339,18 @@ proc ::tincr::catch_info { cmd } {
 #  namespace import ::tincr::*
 interp alias {} ::tincr::import_all {} namespace import ::tincr::*
 
+proc ::tincr::print_quiet { message quiet {newline ""} } {
```
Comment (bradselw) [inline tincr/cad/util/util.tcl]: This is not a good way to handle verbose logging IMO. Tincr already has a global boolean (tincr::debug) that allows the user to turn on debug logging. Perhaps you could use tincr::debug or add another global boolean called tincr::verbose to turn on the verbose logging you added. Then, I would change this method to look something like this:

```
proc ::tincr::print_verbose { message {newline 0} } {
    if {$::tincr::verbose} {
        puts -nonewline $message
        if {$newline} {
            puts {}
        }
    }
}
```

File: tincr/io/design/tincr_checkpoints.tcl [hunk 6]
```diff
@@ -125,84 +125,39 @@ proc ::tincr::read_tcp {args} {
...
-		puts "Linking the design..."
-	}
-	set link_time_string [time {link_design $q -part $part}]
-	set link_runtime [get_time_in_seconds $link_time_string]
-	lappend runtimes $link_runtime
-	if {!$quiet} {
-		puts "Design linked successfully. ($link_runtime seconds)"
-	}
+	::tincr::print_quiet "Reading netlist..." $quiet
+	set edif_runtime [::tincr::time_command "read_edif $q ${filename}/netlist.edf"]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Use `report_runtime` instead of `time_command`

File: tincr/io/design/tincr_checkpoints.tcl [hunk 7]
```diff
@@ -125,84 +125,39 @@ proc ::tincr::read_tcp {args} {
...
+	::tincr::print_quiet "Netlist added successfully. ($edif_runtime seconds)" $quiet
+			
+	::tincr::print_quiet "Linking design..." $quiet
+	set link_runtime [::tincr::time_command "link_design $q -part $part"]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Use `report_runtime` instead of `time_command`

File: tincr/io/design/tincr_checkpoints.tcl [hunk 8]
```diff
@@ -125,84 +125,39 @@ proc ::tincr::read_tcp {args} {
...
+	::tincr::print_quiet "Design linked successfully. ($link_runtime seconds)" $quiet
 	
 	# Disabling placement checks
 	# set placer_checks [get_drc_ruledecks placer_checks]
 	# set_property IS_ENABLED 0 $placer_checks
+		
+	::tincr::print_quiet "Reading design constraints..." $quiet
+	set constraint_runtime [::tincr::time_command "read_xdc $q -no_add ${filename}/constraints.xdc"]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Use `report_runtime` instead of `time_command`

File: tincr/io/design/tincr_checkpoints.tcl [hunk 9]
```diff
@@ -125,84 +125,39 @@ proc ::tincr::read_tcp {args} {
...
-		puts "Reading design constraints..."
-	}
-	set constraint_time_string [time {read_xdc $q -no_add "${filename}/constraints.xdc"}]
-	set constraint_runtime [get_time_in_seconds $constraint_time_string]
-	lappend runtimes $constraint_runtime
-	if {!$quiet} {
-		puts "Design constraints added successfully. ($constraint_runtime seconds)"
-	}
+	::tincr::print_quiet "Reading placement constraints..." $quiet
+	set place_runtime [::tincr::time_command "read_xdc $q -no_add ${filename}/placement.xdc"]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Use `report_runtime` instead of `time_command`

File: tincr/io/design/tincr_checkpoints.tcl [hunk 10]
```diff
@@ -125,84 +125,39 @@ proc ::tincr::read_tcp {args} {
...
-		puts "Reading placement constraints..."
-	}
-	set place_time_string [time {read_xdc $q -no_add "${filename}/placement.xdc"}]
-	set place_runtime [get_time_in_seconds $place_time_string]
-	lappend runtimes $place_runtime
-	if {!$quiet} {
-		puts "Placement constraints added successfully... ($place_runtime seconds)"
-	}
+	::tincr::print_quiet "Reading routing constraints..." $quiet
+	set route_runtime [::tincr::time_command "read_xdc $q -no_add ${filename}/routing.xdc"]
```
Comment (bradselw) [inline tincr/io/design/tincr_checkpoints.tcl]: Use `report_runtime` instead of `time_command`

**Comment by nelsobe** [comment]
Merging this in.  It looks good to me and Thomas.  I need it to continue work on my local changes.  Brad, please review if you have time.

**Comment by bradselw** [comment]
Looks good to me :+1:

**Pull request merged**